### PR TITLE
added option to create private repo

### PR DESCRIPTION
**Pull request by philschmid**
# What does this PR do? 

This PR adds the `repo_visibility` parameter to the `Repository` method to be able to create a private repository when pushing to the hub. 

usage:
```python
    repo = Repository(
        src_dir,
        clone_from=f"{user}/{repository_id}",
        use_auth_token=use_auth_token,
        repo_visibility="private"
    )
    repo.push_to_hub("Adding Infinity model to HF hub")
```

### CODE REVIEW

File: src/huggingface_hub/repository.py [hunk 1]
```diff
@@ -266,6 +266,7 @@ def __init__(
         local_dir: str,
         clone_from: Optional[str] = None,
         repo_type: Optional[str] = None,
+        repo_visibility: Optional[str] = None,
```
Comment (Narsil) [inline src/huggingface_hub/repository.py:269]: Just happene d to have notifications for this repo.

Isn't `repo_visibility: str = "public"` much better than `Optional[str]` ? Strings are immutable so it's fine I think.

File: src/huggingface_hub/repository.py [hunk 2]
```diff
@@ -431,12 +436,14 @@ def clone_from(self, repo_url: str, use_auth_token: Union[bool, str, None] = Non
                 repo_url = repo_url.replace("https://", f"https://user:{token}@")
 
                 if namespace == user or namespace in valid_organisations:
+                    is_repo_private = True if repo_visibility == "private" else False
```
Comment (Narsil) [inline src/huggingface_hub/repository.py:439]: Is that variable a string or a boolean, it seems pretty odd to switch from one to the other.

File: src/huggingface_hub/repository.py (hunk 1 above)
Comment (julien-c) [reply src/huggingface_hub/repository.py:269]: why not use an optional bool instead?

Comment (julien-c) [reply src/huggingface_hub/repository.py:269]: (and name it the same as in `api.create_repo`)

Comment (philschmid) [reply src/huggingface_hub/repository.py:269]: My intention was to have a similar UX and API to what you have when using the UI. There we have repository visibility with `"public"` & `"private"`. But not a strong opinion on this and happy to convert to the same as in `api.create_repo`.

Comment (julien-c) [reply src/huggingface_hub/repository.py:269]: i'll defer to @LysandreJik and/or @osanseviero but personally i think it's better to have a bool

File: src/huggingface_hub/repository.py [hunk 3]
```diff
@@ -404,6 +408,7 @@ def clone_from(self, repo_url: str, use_auth_token: Union[bool, str, None] = Non
         If this folder is a git repository with linked history, will try to update the repository.
         """
         token = use_auth_token if use_auth_token is not None else self.huggingface_token
+        repo_visibility = "public" if token is None else self.repo_visibility
```
Comment (osanseviero) [inline src/huggingface_hub/repository.py:411]: Maybe it's better to throw an error here so users don't accidentally publish a repo publicly.

File: src/huggingface_hub/repository.py (hunk 1 above)
Comment (osanseviero) [reply src/huggingface_hub/repository.py:269]: I think `private: Optional[bool] = None` would be better. That way we avoid having to change from string to bool and we're consistent with `.create_repo`.

File: src/huggingface_hub/repository.py (hunk 3 above)
Comment (philschmid) [reply src/huggingface_hub/repository.py]: Added the error to the top

File: src/huggingface_hub/repository.py [hunk 4]
```diff
@@ -404,6 +408,10 @@ def clone_from(self, repo_url: str, use_auth_token: Union[bool, str, None] = Non
         If this folder is a git repository with linked history, will try to update the repository.
         """
         token = use_auth_token if use_auth_token is not None else self.huggingface_token
+        if token is None and self.private is True:
```
Comment (LysandreJik) [inline src/huggingface_hub/repository.py:411]:
```suggestion
        if token is None and self.private:
```

**Review by LysandreJik — APPROVED** [review]
LGTM

File: src/huggingface_hub/repository.py [hunk 5]
```diff
@@ -404,6 +408,10 @@ def clone_from(self, repo_url: str, use_auth_token: Union[bool, str, None] = Non
         If this folder is a git repository with linked history, will try to update the repository.
         """
         token = use_auth_token if use_auth_token is not None else self.huggingface_token
+        if token is None and self.private:
+            raise ValueError(
+                "You need to provide a Hugging Face Token in `use_auth_token` to work with private repositories"
```
Comment (osanseviero) [inline src/huggingface_hub/repository.py:413]: nit: I think providing the token is more of an edge case for users working in colab or certain scenarios. Maybe we should mention that users should first login with `huggingface-cli login`  or manually provide a token.

**Review by osanseviero — APPROVED** [review]
Looks good! Thanks a lot for this!

File: src/huggingface_hub/repository.py [hunk 6]
```diff
@@ -270,6 +270,7 @@ def __init__(
         git_user: Optional[str] = None,
         git_email: Optional[str] = None,
         revision: Optional[str] = None,
+        private: Optional[bool] = None,
```
Comment (Narsil) [inline src/huggingface_hub/repository.py:273]: Sorry to insist, by `private: bool = False` seems strictly better.

The `None` is necessary when the default argument is mutable, other it's better to use a real default, no ?

**Review by osanseviero — APPROVED** [review]
LG! Thanks

**Pull request merged**